plugins/database/quota/src/org/apache/cloudstack/quota/QuotaServiceImpl.java

```diff
@@ -0,0 +1,304 @@
+                }
+            } else {
+                throw new PermissionDeniedException("Invalid Domain Id or Account");
+            }
+        }
+        TransactionLegacy.open(opendb).close();
+
+        startDate = startDate == null ? new Date() : startDate;
+
+        if (endDate == null) {
+            // adjust start date to end of day as there is no end date
+            Date adjustedStartDate = computeAdjustedTime(_respBldr.startOfNextDay(startDate));
+            if (s_logger.isDebugEnabled()){
+                s_logger.debug("getQuotaBalance1: Getting quota balance records for account: " + accountId + ", domainId: " + domainId + ", on or before " + adjustedStartDate);
+            }
+            List<QuotaBalanceVO> qbrecords = _quotaBalanceDao.lastQuotaBalanceVO(accountId, domainId, adjustedStartDate);
+            if (s_logger.isDebugEnabled()){
+                s_logger.debug("Found records size=" + qbrecords.size());
+            }
+            if (qbrecords.size() == 0) {
+                throw new InvalidParameterValueException("Incorrect Date there are no quota records before this date " + adjustedStartDate);
+            } else {
+                return qbrecords;
+            }
+        } else {
+            Date adjustedStartDate = computeAdjustedTime(startDate);
+            if (endDate.after(_respBldr.startOfNextDay())) {
+                throw new InvalidParameterValueException("Incorrect Date Range. End date:" + endDate + " should not be in future. ");
+            } else if (startDate.before(endDate)) {
+                Date adjustedEndDate = computeAdjustedTime(endDate);
+                if (s_logger.isDebugEnabled()){
+                    s_logger.debug("getQuotaBalance2: Getting quota balance records for account: " + accountId + ", domainId: " + domainId + ", between " + adjustedStartDate + " and " + adjustedEndDate);
+                }
+                List<QuotaBalanceVO> qbrecords = _quotaBalanceDao.findQuotaBalance(accountId, domainId, adjustedStartDate, adjustedEndDate);
+                if (s_logger.isDebugEnabled()){
+                    s_logger.debug("getQuotaBalance3: Found records size=" + qbrecords.size());
+                }
+                if (qbrecords.size() == 0) {
+                    throw new InvalidParameterValueException("Incorrect Date range there are no quota records between these dates start date " + adjustedStartDate + " and end date:" + endDate);
+                } else {
+                    return qbrecords;
+                }
+            } else {
+                throw new InvalidParameterValueException("Incorrect Date Range. Start date: " + startDate + " is after end date:" + endDate);
+            }
+        }
+    }
+
+    @Override
+    public List<QuotaUsageVO> getQuotaUsage(Long accountId, String accountName, Long domainId, Integer usageType, Date startDate, Date endDate) {
```

Comment:
There is appears to be a fair of duplication between this method and ``findQuotaBalanceVO(Long, String, Long, Date, Date)``.  Please extract common elements into private, static methods.  If possible, please add unit tests for the common methods.